src/sparse_blas/backends/mkl_common/mkl_spsv.cxx

```diff
@@ -0,0 +1,180 @@
+    auto internal_A_handle = detail::get_internal_handle(A_handle);
+    if (!internal_A_handle->all_use_buffer()) {
+        detail::throw_incompatible_container(__func__);
+    }
+    if (alg == oneapi::mkl::sparse::spsv_alg::no_optimize_alg) {
+        return;
+    }
+    internal_A_handle->can_be_reset = false;
+    auto event = oneapi::mkl::sparse::optimize_trsv(queue, A_view.uplo_view, opA, A_view.diag_view,
+                                                    internal_A_handle->backend_handle);
+    // spsv_optimize is not asynchronous for buffers as the backend optimize functions don't take buffers.
+    event.wait_and_throw();
+}
+
+sycl::event spsv_optimize(sycl::queue &queue, oneapi::mkl::transpose opA, const void *alpha,
+                          oneapi::mkl::sparse::matrix_view A_view,
+                          oneapi::mkl::sparse::matrix_handle_t A_handle,
+                          oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+                          oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+                          oneapi::mkl::sparse::spsv_alg alg,
+                          oneapi::mkl::sparse::spsv_descr_t /*spsv_descr*/, void * /*workspace*/,
+                          const std::vector<sycl::event> &dependencies) {
+    bool is_alpha_host_accessible = detail::is_ptr_accessible_on_host(queue, alpha);
+    check_valid_spsv(__func__, opA, A_view, A_handle, x_handle, y_handle, is_alpha_host_accessible,
+                     alg);
+    auto internal_A_handle = detail::get_internal_handle(A_handle);
+    if (internal_A_handle->all_use_buffer()) {
+        detail::throw_incompatible_container(__func__);
+    }
+    if (alg == oneapi::mkl::sparse::spsv_alg::no_optimize_alg) {
+        return detail::collapse_dependencies(queue, dependencies);
+    }
+    internal_A_handle->can_be_reset = false;
+    return oneapi::mkl::sparse::optimize_trsv(queue, A_view.uplo_view, opA, A_view.diag_view,
+                                              internal_A_handle->backend_handle, dependencies);
+}
+
+template <typename T>
+sycl::event internal_spsv(sycl::queue &queue, oneapi::mkl::transpose opA, const void *alpha,
+                          oneapi::mkl::sparse::matrix_view A_view,
+                          oneapi::mkl::sparse::matrix_handle_t A_handle,
+                          oneapi::mkl::sparse::dense_vector_handle_t x_handle,
+                          oneapi::mkl::sparse::dense_vector_handle_t y_handle,
+                          oneapi::mkl::sparse::spsv_alg /*alg*/,
+                          oneapi::mkl::sparse::spsv_descr_t /*spsv_descr*/,
+                          const std::vector<sycl::event> &dependencies,
+                          bool is_alpha_host_accessible) {
+    T host_alpha =
+        detail::get_scalar_on_host(queue, static_cast<const T *>(alpha), is_alpha_host_accessible);
+    auto internal_A_handle = detail::get_internal_handle(A_handle);
+    internal_A_handle->can_be_reset = false;
```

Comment:
No I don't. I added this to check to prevent the user from resetting the sparse handle data after it has been used in a computation. We discussed this would be undefined behavior in general when trying to update the sparse handle values or indices without calling the optimize function again.
I'm thinking now that this check I added could be too restrictive since it also prevents users from resetting the sparse handle data after the computation has completed. Maybe I should remove this check since it is "undefined behavior" in the spec. What do you think?

Reply:
Any internal optimizations the backend is making (e.g., generating an internal transposed matrix data) go stale the moment users update values. Regardless of what the spec says, oneMKL backend needs to disallow that at least for now, so this is fine.